lib/UserCohort.js

```diff
@@ -0,0 +1,62 @@
+const mongoose = require("./utils/mongoose");
+const Schema = mongoose.Schema;
+const User = require("./User");
+
+const UserCohort = new Schema(
+  {
+    user: {
+      type: Schema.Types.ObjectId,
+      ref: "User",
+      required: "{PATH} is required!",
+      unique: true
+    },
+    cohort: { type: String }
+  },
+  { timestamps: true }
+);
+UserCohort.plugin(require("./plugins/mongoose-to-id"));
+UserCohort.plugin(require("mongoose-findorcreate"));
+UserCohort.plugin(require("./plugins/mongoose-no-underscore-id"));
+
+UserCohort.statics.findForUser = async function(user) {
+  if (!(user instanceof User)) {
+    throw new Error("user param must be a User instance");
+  }
+  return await this.findOne({ user: user.id }).exec();
+};
+
+UserCohort.statics.setUserCohort = async function(user, cohortName) {
+  if (!(user instanceof User)) {
+    throw new Error("user param must be a User instance");
+  }
+
+  const UserCohortModel = this;
+  let UserCohort = await UserCohortModel.findForUser(user);
+  if (UserCohort && UserCohort.cohort === cohortName) {
+    return UserCohort;
+  }
+
+  UserCohort = await UserCohortModel.findOneAndUpdate(
+    { user: user.id },
+    {
+      user: user.id,
+      cohort: cohortName
+    },
+    { upsert: true, new: true }
+  ).exec();
+
+  // Leave previous cohorts
+  const goalModel = mongoose.model("Goal");
+  const goalCohortModel = mongoose.model("GoalCohort");
+  const goals = await goalModel.find({});
+  for (let i = 0; i < goals.length; i++) {
```

Comment:
this would be way too inefficent at scale (if there were many goals), we should find a way to do this as a single query to `GoalCohort` and then make sure that. that query is properly supported by indexes (I believe it already is)